### DISCUSSION TITLE
Authenticating automated API calls

### DISCUSSION THREAD

**Discussion by demSly** [zitadel/zitadel#1213] [Q&A] [answered]
Our team is developing a REST API, and we are using Zitadel to manage access to it.

We are creating a test system that we want to recreate daily via CI, and create its initial data via Postman collections that are run with Newman.

The issue we are facing is that Postman uses human interaction via browser to get the token needed for the API calls, so we cannot automate initial data creation.

Can you suggest an approach to this? Do you support API keys? I have seen in your Admin Console the "Service Users" but I was not able to create one; can those be a solution to this problem?

**Comment by livio-a** [comment]
Hi @demSly 

We currently don't support API keys in form of static tokens (e.g. personal access token on GitHub, ...)
But as you guessed, your use case can be solved with the mentioned `Service Users`. And it's even more practical, because you can grant permissions to this user just like you would to a "normal" user.

You're basic steps are the following:
1.  create service user
2. create and download key
3. generate JWT (with key)
4. request token with JWT
5. use token for api call

Please check our docs for more info about steps 3 and 4 and  [Using JWTs as Authorization Grants
](https://docs.zitadel.ch/documentation#JSON_Web_Token_JWT_Profile)

You mentioned problem creating the service user. Can you please provide me some more information what the problem was?
Are you still facing this issue?

**Reply by demSly** [reply]
thanks @livio-a 
the problem creating the service user was due to insufficient privileges on my user, that's solved now.
For step 3, I am now writing a command line tool to generate the JWT token using the downloaded key as input. Do you already have something for this purpose?

**Reply by fforootd** [reply]
Hi @demSly 

Currently we have something within our code. 

@livio-a will check if it easily feasible to provide a cli to generate the `jwt`.

**Reply by livio-a** [reply]
@demSly what kind of parameter / input does Newman support?
Would it be helpful to have a cli which takes the key as an input and returns the jwt to stdout or a provided file path?

**Comment by demSly** [comment]
@livio-a what I have in mind is a cli tool that, given the key.json path as input, outputs the JWT token to stdout (or even better, the request token from step 4).

By the way, I tried my hand at writing such a script, and a request with the (expired) token
`curl  --url https://api.zitadel.ch/oauth/v2/token  --header 'Content-Type: application/x-www-form-urlencoded'  --data grant_type=urn:ietf:params:oauth:grant-type:jwt-bearer  --data assertion=[REDACTED] --data scope='openid profile email address'`
returns:
`{"error":"server_error","error_description":"error fetching keys"}`
What is wrong with the token I am generating?

**Reply by fforootd** [reply]
Do you need further assistance?

**Reply by mffap** [reply]
to avoid a potential dead-end, here some more guidance: 

```
POST https://api.zitadel.ch/oauth/v2/token
Content-Type: application/x-www-form-urlencoded

{
  'grant_type': 'urn:ietf:params:oauth:grant-type:jwt-bearer', 
  'scope': 'openid profile urn:zitadel:iam:org:project:id:69234237810729019:aud', 
  'assertion': [jwt from pior step] 
}
```
You must add the scope `urn:zitadel:iam:org:project:id:69234237810729019:aud` without changing this value.

**Comment by livio-a** [comment, marked as answer]
We need the id of the (private) key you signed the JWT with to retrieve the corresponding public key and check the signature of the assertion.
This is done by providing the signer the keyID, which will set the `kid` claim in the JWT Header.

We have now published a cli, which does exactly this: https://github.com/caos/zitadel-tools/releases
